Clarifications to pattern selection (#385)
unicode-org/message-format-wg: spec/formatting.md

When a message has a single _selector_,
an implementation-defined method compares each key to the _selector_
and determines which of the keys match, and in what order of preference.
The list of keys passed to this implementation-defined method
does not include the catch-all key `*`.
The catch-all key `*` is treated as a match for any _selector_,

> but with the lowest possible preference.

Reviewer: Wait... why are we doing this? What is wrong with letting the "implementation-defined method" decide that the `*` key is the best match?

What you're requiring here is that the selector method return "no match" in order to reach `*` and that the higher-level caller of the "implementation-defined method" then choose the fallback vs. allowing the selector method to make that determination. We haven't discussed this previously, so I don't understand why we are changing this text.

Reviewer 2: I don't think this is intended as a change to the selection method, but a clarification of what's already in the Resolve Preferences section below.

If you think that should be changed, that might be best discussed in a separate issue or PR.

Author: @aphillips Yes, as Eemeli said, it's meant to be a clarification of what's already there. The existing definitions of "Resolve Preferences", "Filter Variants", and "Sort Variants" only pass literals to "the implementation-defined method". So the method can do whatever it wants if it's passed anything that's not a literal, such as `*`. Saying "The list of keys passed to this implementation-defined method does not include the catch-all key `*`" is another way of saying that these definitions don't rely on undefined behavior.

Reviewer 2: @aphillips Would you prefer the current proposed change to be included or dropped from this PR? Getting this merged would make it easier for you to file a subsequent PR with your proposed rewrite.

Reviewer: On the one hand, I would prefer not to merge non-consensus text wherever possible. But I agree that the lack of other reviewers leaves little evidence about whether this should or should not go.

Let's merge this. I will file my comments as a separate issue and create a PR with them once this is merged. Sound like a plan?

In a message with more than one _selector_,

> the number of keys in each _variant_ equals the number of _selectors_.

Reviewer: use normative language

```suggestion
the number of keys in each _variant_ **_must_** equal the number of _selectors_.
```

   1. For each _variant_ `var` of the message:
      1. Let `key` be the `var` key at position `i`.
      1. If `key` is not the catch-all key `'*'`:

>          1. Let `ks` be the resolved value of `key`.

Reviewer 2: ```suggestion
         1. Assert that `key` is a _literal_.
         1. Let `ks` be the resolved value of `key`.
```

      1. Let `matchpref` be an integer with the value `minpref`.
      1. Let `key` be the `tuple` _variant_ key at position `i`.
      1. If `key` is not the catch-all key `'*'`:
         1. Assert that `key` is a _literal_.
         1. Let `ks` be the resolved value of `key`.
         1. Let `matchpref` be the integer position of `ks` in `matches`.
      1. Set the `tuple` integer value as `matchpref`.

>    1. Set `sortable` to be `SortVariants(sortable)`.

Reviewer 2: Nitpicky, but let's be a bit more explicit here, and follow the same style of expression as on line 83.

```suggestion
   1. Set `sortable` to be the result of calling the method `SortVariants(sortable)`.
```

   1. Set `i` to be `i` - 1.
1. Let `var` be the _variant_ element of the first element of `sortable`.
1. Select the _pattern_ of `var`.

`SortVariants` is a method whose single argument is
a list of (integer, _variant_) tuples.
It returns a list of (integer, _variant_) tuples.
Any implementation of `SortVariants` is acceptable

> as long as it satisfies the following requirements:
> 1. Let `sortable` be an arbitrary list of (integer, _variant_) tuples.

Reviewer 2: ```suggestion
as long as it satisfies the following requirements:

1. Let `sortable` be an arbitrary list of (integer, _variant_) tuples.
```

The method SortVariants is determined by the implementation.
It takes as an argument a `sortable` list of (integer, _variant_) tuples,
which it modifies in place using some stable sorting algorithm.

> The method does not return a value: `sortable` is its out-parameter.

Reviewer: "out-parameter"?

I actually find this a bit odd or maybe constraining. This defines a method whose output is a side-effect on the input value, which is unusual and it forbids writing a method with a signature like:

```java
public MessageKeyList sortVariants(final MessageKeyList input)
```

Why is it necessary for us to force this?

Reviewer 2: This is intended as a convenience for keeping the method simple. Recall that the section includes this near the top:
> An implementation MAY use any pattern selection method, as long as its observable behaviour matches the results of the method defined here.

On the other hand, if the sense is that it would be easier to read if the implementation-defined method returned some value that was then re-assigned to `sortable`, that's certainly a possible change.

Author: @aphillips The way the algorithm is currently written, `sortable` has to change on each iteration of the `while` loop. However, I think I've found a way to leave it open whether `SortVariants` modifies the list in-place and returns the same list as its result, or returns a new list. Let me know what you think.

Presuming a more powerful implementation which supports selection on numerical values,
and provides a `:plural` function that matches keys by their exact value
as well as their plural category (preferring the former, if possible),
and an English-language formatting context in which
the variable reference `$count` resolves to the number `1`,

> pattern selection proceeds as follows for this message:

Reviewer: Good typo fix.

I find this example nearly impossible to read?

I don't agree with the idea of an "English-language formatting context". It should be in terms of a locale.

I'll come back an suggest an edit when I have a few more minutes.

Author: The edit could maybe be a separate PR?

Reviewer: Actually, on  second thought... it's a typo correct. Fix the typo. We'll address the text separately.
